src/delta/delta.c

```diff
@@ -220,45 +224,45 @@ static int enumerate_dir_d(
         strv_sort(list);
 
         STRV_FOREACH(file, list) {
-                OrderedHashmap *h;
+                OrderedHashmap *h = NULL;
                 int k;
-                char *p;
-                char *d;
+                _cleanup_free_ char *p = NULL, *bp = NULL, *d = NULL;
 
                 if (!endswith(*file, ".conf"))
                         continue;
 
                 p = path_join(path, *file);
                 if (!p)
                         return -ENOMEM;
-                d = p + strlen(toppath) + 1;
+                d = strdup(p + strlen(toppath) + 1);
+                if (!d)
+                        return -ENOMEM;
 
                 log_debug("Adding at top: %s %s %s", d, special_glyph(SPECIAL_GLYPH_ARROW_RIGHT), p);
-                k = ordered_hashmap_put(top, d, p);
+                k = ordered_hashmap_ensure_put(top, &string_hash_ops_free_free, d, p);
                 if (k >= 0) {
                         p = strdup(p);
                         if (!p)
                                 return -ENOMEM;
```

Comment:
there's a memory corruption in this error path, i.e. you use string_hash_ops_free_free now (which is good!), but this means the moment ordered_hashmap_ensure_put() completed destruction of d/p are exclusively the job of the hash table (unless you take the strings out again, but you don't here).

This means if this strdup() fails, then the `_cleanup_free_` for `d` will destroy `d` even though it is also freed by the hashmap. And BOOM! you have a double-free.

basically, once ordered_hashmap_esnure_put() completed you should use TAKE_PTR() on both d and p, to "deactivate" the `_cleanup_` handlers on them. If you want to keep a ref to the strins, then add another local variable (lacking the `_cleanup_` handlers) here and keep a copy of the pointer.

Reply:
I've solved it, but I'm wondering if I could have saved some code by assigning the ref/copy as the return value of `TAKE_PTR`?  I noticed that i.e. `src/network/netdev/macsec.c` did that, but i.e. `src/udev/udev-rules.c` didn't, so I couldn't tell if there was some sort of condition that needs to be satisfied to do that.